File: allergylens_engine/parser.py
"""Turn raw OCR text into a clean list of ingredients.

Handles the messy realities of OCR output: an "Ingredients:" header, line
breaks mid-word, parenthetical sub-ingredients, trailing "Contains:" statements,
and stray punctuation.
"""

from __future__ import annotations

import re
from typing import List

# Where the ingredient list starts and (optionally) ends.
_INGREDIENTS_HEADER = re.compile(r"ingredients?\s*[:\-]", re.IGNORECASE)
_END_MARKERS = re.compile(
    r"\b(contains|may contain|manufactured|distributed|allergen|nutrition|"
    r"net wt|best by|directions|storage)\b",
    re.IGNORECASE,
)


def find_ingredients_section(raw_text: str) -> str:
    """Return the substring that holds the ingredient list.

    Falls back to the whole text if no explicit "Ingredients:" header is found,
    so free-form pasted lists still parse.
    """
    if not raw_text:
        return ""

    header = _INGREDIENTS_HEADER.search(raw_text)
    section = raw_text[header.end():] if header else raw_text

    end = _END_MARKERS.search(section)
    if end:
        section = section[: end.start()]

    return section.strip()


def clean_text(raw_text: str) -> str:
    """Normalize whitespace and repair common OCR line-break artifacts."""
    text = raw_text.replace("\r", "\n")
    # Join words split across a line break by a hyphen: "leci-\nthin" -> "lecithin"
    text = re.sub(r"-\s*\n\s*", "", text)
    # Treat remaining newlines as spaces.
    text = re.sub(r"\s*\n\s*", " ", text)
    # Collapse runs of whitespace.
    text = re.sub(r"\s{2,}", " ", text)
    return text.strip()


def _clean_ingredient(item: str) -> str:
    item = item.strip().strip(".;:*•·-")
    # Drop parenthetical sub-ingredient qualifiers, e.g. "flour (wheat)" -> "flour".
    item = re.sub(r"\([^)]*\)", "", item)
    item = re.sub(r"\s{2,}", " ", item).strip()
    return item.lower()


def parse_ingredients(raw_text: str) -> List[str]:
    """Parse raw OCR/label text into a de-duplicated list of ingredients."""
    section = find_ingredients_section(raw_text)
    section = clean_text(section)
    if not section:
        return []

    # Split on the usual separators between ingredients.
    parts = re.split(r"[,;•·\|]| and ", section)

    seen = set()
    ingredients: List[str] = []
    for part in parts:
        cleaned = _clean_ingredient(part)
        if not cleaned or len(cleaned) < 2:
            continue
        if cleaned in seen:
            continue
        seen.add(cleaned)
        ingredients.append(cleaned)

    return ingredients

File: web/lib/parser.ts
// Turn raw OCR/label text into a clean ingredient list.
// Mirrors allergylens_engine/parser.py.

const INGREDIENTS_HEADER = /ingredients?\s*[:\-]/i;
const END_MARKERS =
  /\b(contains|may contain|manufactured|distributed|allergen|nutrition|net wt|best by|directions|storage)\b/i;

/** Return the substring holding the ingredient list (whole text if no header). */
export function findIngredientsSection(rawText: string): string {
  if (!rawText) return "";

  const header = INGREDIENTS_HEADER.exec(rawText);
  let section = header ? rawText.slice(header.index + header[0].length) : rawText;

  const end = END_MARKERS.exec(section);
  if (end) section = section.slice(0, end.index);

  return section.trim();
}

/** Normalize whitespace and repair common OCR line-break artifacts. */
export function cleanText(rawText: string): string {
  let text = rawText.replace(/\r/g, "\n");
  // Join words split across a line break by a hyphen: "leci-\nthin" -> "lecithin".
  text = text.replace(/-\s*\n\s*/g, "");
  // Treat remaining newlines as spaces.
  text = text.replace(/\s*\n\s*/g, " ");
  // Collapse runs of whitespace.
  text = text.replace(/\s{2,}/g, " ");
  return text.trim();
}

function cleanIngredient(item: string): string {
  let out = item.trim().replace(/^[.;:*•·\-\s]+|[.;:*•·\-\s]+$/g, "");
  out = out.replace(/\s{2,}/g, " ").trim();
  return out.toLowerCase();
}

/** Parse raw text into a de-duplicated ingredient list. */
export function parseIngredients(rawText: string): string[] {
  let section = cleanText(findIngredientsSection(rawText));
  if (!section) return [];

  // Flatten parenthetical sub-ingredients so hidden allergens aren't lost:
  // "chocolate (cocoa, milk fat)" -> "chocolate, cocoa, milk fat".
  section = section.replace(/[()[\]]/g, ", ");

  const parts = section.split(/[,;•·|]| and /);

  const seen = new Set<string>();
  const ingredients: string[] = [];
  for (const part of parts) {
    const cleaned = cleanIngredient(part);
    if (!cleaned || cleaned.length < 2 || seen.has(cleaned)) continue;
    seen.add(cleaned);
    ingredients.push(cleaned);
  }
  return ingredients;
}
